src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -0,0 +1,264 @@
+                }
+            }
+
+            if ((W3CLoggingFields.Request & options.LoggingFields) != W3CLoggingFields.None)
+            {
+                var request = context.Request;
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.ProtocolVersion))
+                {
+                    AddToList(elements, W3CLoggingFields.ProtocolVersion, request.Protocol);
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.Method))
+                {
+                    AddToList(elements, W3CLoggingFields.Method, request.Method);
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.UriStem))
+                {
+                    AddToList(elements, W3CLoggingFields.UriStem, request.Path.Value);
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.UriQuery))
+                {
+                    AddToList(elements, W3CLoggingFields.UriQuery, request.QueryString.Value);
+                }
+
+                if ((W3CLoggingFields.RequestHeaders & options.LoggingFields) != W3CLoggingFields.None)
+                {
+                    var headers = request.Headers;
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.Host))
+                    {
+                        if (headers.TryGetValue(HeaderNames.Host, out var host))
+                        {
+                            AddToList(elements, W3CLoggingFields.Host, host.ToString());
+                        }
+                    }
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.Referer))
+                    {
+                        if (headers.TryGetValue(HeaderNames.Referer, out var referer))
+                        {
+                            AddToList(elements, W3CLoggingFields.Referer, referer.ToString());
+                        }
+                    }
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.UserAgent))
+                    {
+                        if (headers.TryGetValue(HeaderNames.UserAgent, out var agent))
+                        {
+                            AddToList(elements, W3CLoggingFields.UserAgent, agent.ToString());
+                        }
+                    }
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.Cookie))
+                    {
+                        if (headers.TryGetValue(HeaderNames.Cookie, out var cookie))
+                        {
+                            AddToList(elements, W3CLoggingFields.Cookie, cookie.ToString());
+                        }
+                    }
+                }
+            }
+
+            var response = context.Response;
+
+            try
+            {
+                await _next(context);
+            }
+            catch
+            {
+                // Write the log
+                if (_hasLogged)
```

Comment:
I don't understand the meaning of this variable. It seems to mean that some log fields are turned on and have been copied to the elements array, but those haven't been written to the logger? Would `_shouldLog` capture that better?

Reply:
Wait, _hasLogged is a class variable but is being updated per request? That's a bug, you're mixing state across requests. This needs to be tracked per request as a ref or return value from AddToList

Reply:
Good catch, should be per-request. It's to indicate whether there's actually anything to log in this request (e.g. if the user turned on only the "cookie" field, but there was no cookie in the HttpContext, there'd be no use in calling W3CLogger.Log())

Reply:
And yeah, `_shouldLog` seems clearer